File: forcomp.scala
package forcomp
import common._

object Anagrams {
  type Word = String
  type Sentence = List[Word]
  type Occurrences = List[(Char, Int)]

  val dictionary: List[Word] = loadDictionary

  // ex. 1a
  def wordOccurrences(w: Word): Occurrences = {
    w.toLowerCase.groupBy(identity).map((c) => (c._1, c._2.length)).toList.sortBy(_._1)
  }

  // ex. 1b
  def sentenceOccurrences(s: Sentence): Occurrences = wordOccurrences(s.foldRight("")(_++_))

  // ex. 2a
  lazy val dictionaryByOccurrences: Map[Occurrences, List[Word]] = {
    dictionary.groupBy(wordOccurrences)
  }

  // ex. 2b
  def anagramsByOccurrences(occurrences: Occurrences): List[Word] = {
    val all = dictionaryByOccurrences.get(occurrences)
    if (all.isEmpty) Nil else all.get
  }
  def wordAnagrams(word: Word): List[Word] = anagramsByOccurrences(wordOccurrences(word))

  // ex. 3
  def combinations(occurrences: Occurrences): List[Occurrences] = {
    if (occurrences.isEmpty)
      List(Nil)
    else {
      val next = occurrences.head
      val rest = combinations(occurrences.tail)
      val result = for (i <- 0 to next._2; subset <- rest) yield {
        if (i == 0) subset else (next._1, i) :: subset
      }
      result.toList
    }
  }

  // ex. 4a
  def subtract(x: Occurrences, y: Occurrences): Occurrences = {
    def updateMap(a: Map[Char, Int], b: (Char, Int)) = {
      val n = a.get(b._1).get // we know that this exists
      if (n == b._2)
        a - b._1
      else
        a.updated(b._1, n - b._2)
    }
    y.foldLeft(x.toMap)(updateMap).toList.sortBy(_._1)
  }

  // ex. 4b
  def sentenceAnagrams(sentence: Sentence): List[Sentence] = {
    def rec(occurrences: Occurrences): List[Sentence] = {
      if (occurrences.isEmpty)
        List(Nil)
      else {
        for (comb <- combinations(occurrences);
             word <- anagramsByOccurrences(comb);
             sentence <- rec(subtract(occurrences, comb)))
        yield word :: sentence
      }
    }
    rec(sentenceOccurrences(sentence))
  }

}

File: forcomp.hs
import Control.Arrow ((&&&))
import Control.Monad.Reader
import Data.Char (toLower)
import Data.Function (on)
import Data.List (group, groupBy, sort, sortBy)
import Data.Map (Map)
import qualified Data.Map as Map

type Word        = String
type Sentence    = [Word]
type Occurrences = [(Char, Int)]
type Dictionary  = Map Occurrences [Word]

-- same functionality as Scala's groupBy - can we do this faster?
groupIntoMapBy f = Map.fromList . map (f . head &&& id)
                   . groupBy ((==) `on` f) . sortBy (compare `on` f)

wordOccurrences         = map (head &&& length) . group . sort . map toLower
sentenceOccurrences     = wordOccurrences . foldr (++) ""
dictionaryByOccurrences = groupIntoMapBy wordOccurrences
anagramsByOccurrences :: Occurrences -> Reader Dictionary [Word]
anagramsByOccurrences x = asks (Map.lookup x) >>= return . (maybe [] id)
wordAnagrams            = anagramsByOccurrences . wordOccurrences

combinations []     = [[]]
combinations (x:xs) = [prefixIfNotZero i rest | i <- [0..snd x], rest <- combinations xs]
  where prefixIfNotZero i rest = if i == 0 then rest else (fst x, i) : rest

subtractOccurrences x y = convertBack $ Map.differenceWith diff (Map.fromList x) (Map.fromList y)
  where diff i j    = if i == j then Nothing else Just (i - j)
        convertBack = sortBy (compare `on` fst) . Map.toList

sentenceAnagrams :: Sentence -> Reader Dictionary [Sentence]
sentenceAnagrams = rec . sentenceOccurrences
  where rec     []     = return [[]]
        rec     xs     = fmap concat $ sequence [process xs occ | occ <- combinations xs]
        process xs occ = do rest  <- rec $ subtractOccurrences xs occ
                            words <- anagramsByOccurrences occ
                            return [w : r | w <- words, r <- rest]

-- Test

dictFile = "linuxwords.txt"

main = do dic <- fmap (dictionaryByOccurrences . lines) $ readFile dictFile
          mapM_ (putStrLn . show) $ runReader (sentenceAnagrams ["yes", "man"]) dic
